docs/enhancements/automatic-agent-classification-labels.md

```diff
@@ -0,0 +1,101 @@
+
+This proposal is to automate user-defined labels.  Users would define the desired labels and conditions, and the system would automate the process.
+
+## Motivation
+
+Having users define labels manually is time-consuming and error-prone.  Defining conditions is more reliable and scalable for users.
+
+### Goals
+
+- Provide a flexible and easy-to-use API for defining desired labels and conditions.
+- Extend the Infrastructure Operator to maintain the desired labels based on the specified conditions with minimal impact on scale.
+
+### Non-Goals
+
+- Ship pre-defined labels and conditions.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Cluster Creator, I want the system to automatically label Agents based on specific conditions so that I can more easily select Agents for cluster membership.	
+
+## Design Details [optional]
+
+We will create a new AgentClassification namespace-scoped CRD that defines the desired label key-value and a list of expressions to be evaluated according to the Agent’s inventory.  Each expression is generic such that it can refer to any properties in the inventory.
+
+### AgentClassification CRD Definition
+
+```
+// AgentClassificationSpec defines the desired state of AgentClassification
+type AgentClassificationSpec struct {
+	// LabelKey specifies the label key to apply to matched Agents
+	//
+	// +immutable
+	LabelKey string `json:"labelKey"`
+
+	// LabelValue specifies the label value to apply to matched Agents
+	//
+	// +immutable
+	LabelValue string `json:"labelValue"`
+
+	// ExpressionDefinitions define the conditions upon which the specified label is applied
+	ExpressionDefinitions []ExpressionDefinition `json:"expressionDefinitions"`
+}
+
+// ExpressionDefinition defines an expression to be evaluated
+type ExpressionDefinition struct {
+	// Expression is the expression that will be evaluated, in govaluate format
+	Expression string `json:"expression"`
+
+	// Parameters maps parameters from the above expression to paths in the Agent inventory
+	Parameters map[string]string `json:"parameters"`
+}
+```
+
+We will base the expression definition on the [govaluate](https://github.com/Knetic/govaluate) library, which supports a language for expressions with parameters.  Along with each expression (e.g., “cpu > 8”), the user will define a list of variables and map them to inventory items (e.g., “cpu”: “Cpu.Count”).  The controller will need to parse the inventory path and translate it into the corresponding value.  In addition, the inventory contains arrays for disks and network interfaces - the user will pass "Disks(N).SizeBytes" for at least N disks having the specified capacity, or "Disks().SizeBytes" for all disks.  The controller will need to parse this and create additional govaluate expressions for each disk, and then check for at least N returning true.
```

Comment:
>"Disks(N).SizeBytes" for at least N disks having the specified capacity, or "Disks().SizeBytes" for all disks

This sounds pretty limiting.

For example, we have disks in the inventory which are not fit for installation. Usually these disks you'd never want to use at all, not just for installation. Users would probably want to say things like "all disks that are eligible for installation" or "all SSD disks", i.e. do some sort of filtering on the network/disk arrays, and this DSL would not allow such flexibility.

Could we somehow make use of an off-the-shelf JSON query language, something like the `jq` query language, or jsonpath?

Reply:
i see that there something like that https://github.com/itchyny/gojq

Reply:
Great suggestion, I will look into using this instead of govaluate.